### PR TITLE
Add explanation for decision tree

### PR DESCRIPTION
**Pull request by xiaoye-hua**
@bfortuner  Hi. Brendan. I've add  an explanation of decision tree including ID3，C4.5 and CART. Please help me to review and merge these changes.

### CODE REVIEW

File: docs/classification_algos.rst [hunk 1]
```diff
@@ -11,21 +11,152 @@ Bayesian
...
+1. Assign all training instances to the root of the tree. Set current node to root node.
+2. Find the split feature and split value based on the split criterion such as information gain, information gain ratio or gini coefficient.
+3. Partition all data instances at the node based on the split feature and threshold value.
+4. Denote each partition as a child node of the current node.
+5. For each child node:
+    1. If the child node is “pure” (has instances from only one class), tag it as a leaf and return.
+    2. Else, set the child node as the current node and recurse to step 2.
+
+
+ID3 creates a multiway tree. For each node, it try to find the categorical feature that will yield the largest information gain for the target variable.
```
Comment (ivanistheone) [inline docs/classification_algos.rst:40]: it trie**s**

File: docs/classification_algos.rst [hunk 2]
```diff
@@ -11,21 +11,152 @@ Bayesian
...
+
+C4.5 is the successor of ID3 and remove the restriction that the feature must be categorical by dynamically define a discrete attribute that partitions the continuous attribute in the discrete set of intervals.
+
+CART is similar to C4.5. But it differs in that it constructs binary tree and support regression problem [#sklearntree]_.
+
+The main differences are shown in the follow table:
```
Comment (ivanistheone) [inline docs/classification_algos.rst:46]: follow --> following

File: docs/classification_algos.rst [hunk 3]
```diff
@@ -11,21 +11,152 @@ Bayesian
...
+                else:
+                    # Split the node
+                    child_nodes = self.split_node(node)
+                    if not child_nodes:
+                        self.set_label(node)
+                    else:
+                        queue.extend(child_nodes)
+..
+
+For CART algorithm, when constructing the binary tree, it will try search for the feature and threshold that will yield the largest gain or the least impurity. The split criterion is a combination of the child nodes' impurity. For the child nodes' impurity, gini coefficient or information gain are adopted in classification. For regression problem, mean-square-error or mean-absolute-error are used. Example codes are showed below. For more details about the formulas, please refer to `Mathematical formulation for decision tree in scikit-learn documentation <https://scikit-learn.org/stable/modules/tree.html#mathematical-formulation>`__
```
Comment (ivanistheone) [inline docs/classification_algos.rst:116]: For CART algorithm, when constructing the binary tree, **it** will try search for
-->
**The** CART algorithm, when constructing the binary tree, will try search**ing** for

**Comment by ivanistheone** [comment]
seems to be also a .pyc file not needed, see `code/decision_tree/__pycache__/decision_tree_base.cpython-36.pyc`

**Review by ivanistheone — CHANGES_REQUESTED** [review]
I made some minor suggestions for typo fixes. Code look good, though I haven't tried running it.

**Comment by xiaoye-hua** [comment]
> I made some minor suggestions for typo fixes. Code look good, though I haven't tried running it.

@ivanistheone Hi. Thank you for your review. I've correct these typo errors and removed the unrelated file.

**Pull request merged**